design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,353 @@
+3. Synchronize the labels and update the Node in the workload cluster.
+
+An alternative would be to set up a watch on the Node resource types in every
+workload cluster. The benefit being that we can quickly synchronize any label
+changes on the Node object itself. However, this introduces complications when
+the management cluster != workload cluster and additionally, when there are
+many workload clusters being managed by a single management cluster. How such
+a watch could be implemented within the controller-runtime model would require
+further investigation.
+
+### Configuration
+
+#### Command-Line Flag
+
+We need to expose two additional configuration variables to the user: the set
+of label prefixes and the full-sync interval. We can introduce two new
+command-line flags for CAPM3, and also utilize default values if the flags are
+not specified: (i) `--bmh-label-prefixes` with a nil default value and (ii)
+`--label-sync-interval` with a 15-second default value.
+
+Unfortunately, command-line flags will not be compatible with openShift where
+cluster administrators do not deploy metal3 themselves and do not have access
+to the command line arguments for the cluster-api provider.
+
+#### Annotations on BMH
+
+In this approach, an annotation is added to a BMH, in addition to the labels
+to be synchronized, which reflects the prefixes to be utilized. For example,
+
+```yaml
+  kind: BareMetalHost
+  metadata:
+    annotations:
+      cluster.x-k8s.io/label-prefixes: "my-prefix.metal3.io, my-prefix.kubernetes.io"
+      cluster.x-k8s.io/label-sync-interval: "30s"
+    labels:
+      my-prefix.metal3.io/rack: xyz-123
+      my-prefix.kubernetes.io/zone: security-level-0
+```
+
+A concern with this approach is consistency since the same annotation would
+have to be duplicated across BMHs. The prefixes, while user configurable, are
+static and should not require re-configuring. For example, in a likely
+scenario, the cluster admin would define a standard set of prefixes to
+be used internally. Also, annotations can be (accidentally) added/removed/
+changed, leading to undesirable behavior.
+
+#### New API Type
+
+We could introduce a new API type: `BareMetalHostLabelSyncProfile` (any other
+suggestions for the name?).
+
+The common case is when a label on the BMH is updated, we must first find the
+corresponding `BareMetalHostLabelSyncProfile` (if any) that matches the BMH.
+To avoid fetching all `BareMetalHostLabelSyncProfile`, we can make use of a
+label on the BMH, in namespace/name format, that references the
+`BareMetalHostLabelSyncProfile` (see example below). We can drop the namespace
+requirement if it's in the same namespace as the BMH.
+
+The less common case is where the `BareMetalHostLabelSyncProfile` is updated (
+e.g. a prefix is removed). In this case, we will need to fetch all BMHs
+associated and perform a sync operations across the impacted Nodes. Label
+selectors could be used for this purpose.
```

Comment:
This sounds a lot like what the hardware-classification-controller had to do.

I think it's going to be a lot easier to get the syncing logic right if we can figure out a way to watch resources in a "normal" way instead of trying to batch the changes. Maybe you could elaborate more on why that's not possible? I understand why we can't watch Nodes, since they may be in a separate cluster. Why can't we watch hosts? Or why can't we leave it up to each cluster to manage the labels for its own hosts & nodes?

Reply:
> I think it's going to be a lot easier to get the syncing logic right if we can figure out a way to watch resources in a "normal" way instead of trying to batch the changes. Maybe you could elaborate more on why that's not possible? 

The controller watches the BareMetalHostLabelSyncProfile for changes (e.g. prefix updates). However, the profile is associated with 1 or more BMHs. In order to retrieve these BMHs, we make use of labels. I don't see where we are batching the changes. 
We do watch BMHs as discussed in the paragraph above this one, but that is for changes to the BMHs itself. 
If there is another way, please let me know. 

> I understand why we can't watch Nodes, since they may be in a separate cluster. Why can't we watch hosts? Or why can't we leave it up to each cluster to manage the labels for its own hosts & nodes?

I'm not sure I fully understand your comment. Are you suggesting another controller running in the workload cluster that watches BMHs in the mgmt cluster? In general, the pattern of a workload cluster initiating a call to the mgmt cluster introduces more complexities. For example, (i) placing appropriate credentials in the workload cluster, (ii) what happens when the mgmt cluster is something like KinD which is not externally visible, .... The CAPI community generally frowns on this pattern as well.

Reply:
Where is the controller that watches `BareMetalHostLabelSyncProfile` going to run? How is it going to find/modify Nodes? Where are the hosts that it will read to know which labels to apply to those nodes?

Reply:
The controller runs on the management cluster and is part of CAPM3. If you recall, this new api type is an alternative to using command-line flags to pass the prefixes (which would not work for openshift). Nothing else has changed.

Also, please see the below section on using Metal3Cluster as an alternative. There was some discussion on that with @maelk. The discussion can still be found under the Comments tab.

Reply:
OK, I suggested watching hosts and nodes as well as the new sync profile resource. You said that wasn't possible because those resources might not be in the same cluster as the profile, so I was confused. If that controller isn't in the same place as the Node it is going to change, how does it change the Node's labels? This is probably obvious to you, but my distro uses these components in a different way (there is no cluster pivot, everything is always just in the same cluster) so I apologize if this is a really basic question.

Reply:
Sorry about the confusion. 

If the mgmt cluster == workload cluster, the process likely simplified. Unfortunately, we can' make that assumption. The process for updating the Nodes in the workload cluster is described [here in this proposal](https://github.com/metal3-io/metal3-docs/pull/149/files#diff-413f1e6764de1ed31caea7e636db03eaea13a1bbd33d2b4a2b52ef507237718cR110). CAPI maintains the credentials for accessing each workload cluster inside a Secret. There is precedence for fetching/updating resources in the workload cluster in both capm3 and capi, so we would take the same approach here.